pkg/components/linkerd/component.go

```diff
@@ -46,12 +46,24 @@ type component struct {
 	Cert   string
 	Key    string
 	Expiry string
+
+	metadata *components.Metadata
 }
 
 func newComponent() *component {
 	return &component{
 		ControllerReplicas: 1,
 		EnableMonitoring:   false,
+
+		metadata: &components.Metadata{
```

Comment:
Configuration and metadata should be separated, as their origins are different. Configuration is specified by the user while metadata is defined by the author of the component.

Reply:
Should I define a package global Metadata?

Reply:
No, as this is the struct (so it can be const) and we should not use globals, IMO `Metadata()` function is the good place to put this. Or do you mean something else?

Reply:
I want the labels and annotations to be loaded automatically and not hardcoded in the go code. What is the ideal way to do it? Current implementation relies on the developer making upgrades to not forget adding annotations or labels as they are updated in the component.

Reply:
IMO it should be fine to put them into Go code at the moment. And yes, developer must remember to update them when doing the upgrades. It is also the case for other projects as far as I know, like `cert-manager`.

Reply:
Yes that's what I am trying to avoid by making it load dynamically.

Reply:
I think the complexity of avoid it outweighs the benefit of having it loaded dynamically. Also, other charts do not include namespace objects, so the dynamic load will be only done by a single component anyway.

Reply:
I can try to reduce complexity if you point out specific instances of it.

Right now I haven't touched other components but with this change other components can be updated accordingly. Like istio, cert-manager, etc.

Reply:
- cert-manager tells you to create the namespace manually, not via manifest in the chart: https://cert-manager.io/docs/installation/kubernetes/#steps
- istio recommends the same: https://istio.io/v1.3/docs/setup/install/helm/#option-1-install-with-helm-via-helm-template

It seems `linkerd2` chart doesn't follow Helm charts best practices, so I wouldn't make our code compatible with just this one special chart, which is experimental anyway.

Reply:
This makes sense. If we are adding the namespace config by modifying upstream then we might as well maintain that in the code and not in the helm chart.